File: Two Sum - Difference equals to target/Javascript/solution_5.js
// https://www.lintcode.com/problem/610/

// 采用同向双指针的解法
function twoSum7(nums, target) {
  let len = nums.length;
  target = Math.abs(target);

  let j = 1;

  for (let i = 0; i < len; i++) {
    j = Math.max(j, i + 1);
    while (nums[j] - nums[i] < target) {
      j++;
    }

    if (j >= len) {
      break;
    }

    if (nums[j] - nums[i] === target) {
      return [nums[i], nums[j]];
    }
  }

  return [-1, -1];
}

let nums = [0, 5, 7];
let target = -2;
console.log(twoSum7(nums, target));

File: Two Sum - Difference equals to target/Rust/src/solution_5.rs
// https://www.lintcode.com/problem/610/

use std::cmp::max;

pub fn two_sum7(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let nums_len = nums.len();
    let target = target.abs();
    let mut result = vec![];

    let mut j = 1;

    for i in 0..nums_len {
        j = max(j, i + 1);

        while nums[j] - nums[i] < target {
            j += 1;
        }

        if j >= nums_len {
            break;
        }

        if nums[j] - nums[i] == target {
            return vec![nums[i], nums[j]];
        }
    }

    result
}